src/gevent/pool.py

```diff
@@ -714,13 +722,30 @@ def free_count(self):
             return 1
         return max(0, self.size - len(self))
 
-    def add(self, greenlet):
+    def add(self, greenlet, blocking=True, timeout=None):
         """
-        Begin tracking the given greenlet, blocking until space is available.
+        Begin tracking the given greenlet, possibly blocking until space is
+        available.
+
+        :keyword bool blocking: If True (the default), this function will block
+        until the pool has space or a timeout occurs. If False, this function
+        will immediately raise a Timeout if the pool is currently full.
+        :keyword float timeout: The maximum number of seconds this method will
+        block, if ``blocking`` is True. (Ignored if ``blocking`` is False.)
+
+        Raises ``Timeout`` on timeout, or `Full` if ``blocking`` is False and
+        the pool is full.
 
         .. seealso:: :meth:`Group.add`
+
+        .. versionchanged:: 1.3.0
+            Added the ``blocking`` and ```timeout`` parameters.
         """
-        self._semaphore.acquire()
+        if not self._semaphore.acquire(blocking=blocking, timeout=timeout):
+            # We failed to acquire the semaphore. Presumably, blocking was False, because had it
+            # been True, we would have either acquired the semaphore or encountered a Timeout.
```

Comment:
The comment is incorrect. The Semaphore will never itself raise a Timeout, *regardless* of the `blocking` parameter. That's the cause of [this test failure](https://travis-ci.org/gevent/gevent/jobs/299331466#L1970)

When you acquire a semaphore, if it is available you immediately get it (it returns True). If it's not available and `blocking` was False, the semaphore immediately returns False. The only remaining case is that it wasn't available, and blocking was True; in that case, the semaphore will wait until it gets it and return True (or, if you specified a timeout that elapses before that happens, return False).

Reply:
Addressed.